src/backends/torch/optim/ranger.cc

```diff
@@ -227,11 +237,40 @@ namespace dd
                 slow_p.add_(p.data() - slow_p, options.lalpha());
                 p.data().copy_(slow_p);
               }
+
+            if (options.swa())
+              {
+                auto &swa_buf = state.swa_buffer();
+                double swa_decay = 1.0 / (state.step() + 1);
+                torch::Tensor diff = (p.data() - swa_buf) * swa_decay;
+                swa_buf.add_(diff);
+              }
           }
       }
     return loss;
   }
 
+  void Ranger::swap_swa_sgd()
+  {
+    for (auto &group : param_groups_)
+      {
+        for (auto &p : group.params())
+          {
+            auto &options = static_cast<RangerOptions &>(group.options());
+            if (!options.swa())
+              continue;
```

Comment:
Move in outer loop, since the whole group can be skipped?

Reply:
ok, done